### ISSUE TITLE
Question: Using Grad(F,V,E) where F is a PyTorch Module

### ISSUE THREAD

**Issue by EdwardRaff** [getkeops/keops#72]
Title has the question. To simplify, I want to be able to perform a reduction over a function F : R^d -> scalar . and its gradient ∇F. e.g. I want to reduce over a function similar to `F(x) + ∇F(x)`. 

I see that KerOps has a `Grad(F,V,E)` function, so hypothetically I could have something like `F(x)+ Grad(F,x, x)`. My problem is that I want F to be a neural network itself, and to learn it's parameters as part of this process. 

Is that something that can be done in KeOps?

**Comment by jeanfeydy** [comment]
Hi @EdwardRaff ,

Indeed, this should be perfectly do-able! 
(All my apologies for the late reply: we've been extremely busy over the last two months but are now catching up on late issues.)

KeOps supports matrix-vector products within formulas `F(x)` (e.g. the [`.matvecmult` LazyTensor method](http://www.kernel-operations.io/keops/python/api/pytorch/LazyTensor.html#pykeops.torch.LazyTensor.matvecmult)), and we routinely use [multi-layer perceptrons](https://en.wikipedia.org/wiki/Multilayer_perceptron) within KeOps reductions. Backpropagation is fully supported even with higher-order derivatives, so you won't have any problem learning neural weights using standard PyTorch optimizers. 
If you include a small working example of the network you intend to work with, I could probably give you more details on the precise syntax.

Best regards,
Jean

**Comment by jeanfeydy** [comment]
I'm closing this issue for now: feel free to re-open if needed!
